openstack/client.go

```diff
@@ -212,125 +215,81 @@ func NewIdentityV3(client *gophercloud.ProviderClient, eo gophercloud.EndpointOp
 	return &gophercloud.ServiceClient{
 		ProviderClient: client,
 		Endpoint:       endpoint,
+		Type:           clientType,
 	}, nil
 }
 
-// NewObjectStorageV1 creates a ServiceClient that may be used with the v1 object storage package.
-func NewObjectStorageV1(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (*gophercloud.ServiceClient, error) {
-	eo.ApplyDefaults("object-store")
+func initClientOpts(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts, clientType string) (*gophercloud.ServiceClient, error) {
+	sc := new(gophercloud.ServiceClient)
+	eo.ApplyDefaults(clientType)
 	url, err := client.EndpointLocator(eo)
 	if err != nil {
-		return nil, err
+		return sc, err
 	}
-	return &gophercloud.ServiceClient{ProviderClient: client, Endpoint: url}, nil
+	sc.ProviderClient = client
+	sc.Endpoint = url
+	sc.Type = clientType
```

Comment:
It is assumed that `client.EndpointLocator` will return an endpoint URL that's service type equals the requested `clientType`. I think this is fine and it's how `client.EndpointLocator` works now.
However, there are other service information that are available in [`func V2EndpointURL`](https://github.com/gophercloud/gophercloud/blob/master/openstack/endpoint_location.go#L15) and [`func V3EndpointURL`](https://github.com/gophercloud/gophercloud/blob/master/openstack/endpoint_location.go#L63) and are not returned by the functions and that may be useful in the `gophercloud.ServiceClient` struct in the future.
IMHO, we may open the doors for returning these information from the functions [by adding one more return value to the `EndpointLocator` type](https://github.com/gophercloud/gophercloud/pull/352/commits/e8c06e99db6f207b542e75bba521e2b8a978cc84).